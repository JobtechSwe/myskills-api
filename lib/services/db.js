const consents = new Map()
const consentRequests = new Map()

// TODO: We should really store this in a DB
module.exports = {
    saveConsent: consent => consents.set(consent.consentId, consent),
    getConsent: id => consents.get(id),
    saveConsentRequest: consent =>
        consentRequests.set(consent.consentRequestId, consent),
    getConsentRequest: id => consentRequests.get(id),
}
