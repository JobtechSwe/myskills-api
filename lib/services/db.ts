import redis from '../adapters/redis'

const consentPrefix = 'consentId:'
const consentRequestPrefix = 'consentRequestId:'

export const saveConsent = (consent: any) =>
  redis.set(`${consentPrefix}${consent.consentId}`, JSON.stringify(consent))

export const getConsent = async (id: string) => {
  const data = await redis.get(`${consentPrefix}${id}`)
  return JSON.parse(data)
}

export const saveConsentRequest = (consent: any) =>
  redis.set(
    `${consentRequestPrefix}${consent.consentRequestId}`,
    JSON.stringify(consent)
  )

export const getConsentRequest = async (id: string) => {
  const data = await redis.get(`${consentRequestPrefix}${id}`)
  return JSON.parse(data)
}
