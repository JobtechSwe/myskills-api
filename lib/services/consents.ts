import config from '../config'
import { Area, SubscriptionMessage } from '../types'
import { saveConsent, saveConsentRequest } from './db'
import { Consent, Login } from '@mydata/client'
import pubSub from '../adapters/pubsub'

export interface Scope {
  area: Area
  description: string
  domain: string
  lawfulBasis: string
  permissions: string[]
  purpose: string
}

export interface DefaultRequest {
  expiry: number
  scope: Scope[]
}

const createConsent = (area: Area) => ({
  area,
  description: `${area} that you have entered in the service.`,
  domain: config.DOMAIN,
  lawfulBasis: 'CONSENT',
  permissions: ['write'],
  purpose: 'In order to create a CV using our website.',
})

const defaultRequest = (durationInSeconds: number): DefaultRequest => ({
  expiry: Math.round(Date.now() / 1000 + durationInSeconds),
  scope: [
    createConsent(Area.languages),
    createConsent(Area.educations),
    createConsent(Area.experiences),
    createConsent(Area.skills),
    createConsent(Area.profile),
  ],
})

const onConsentApproved = async (payload: Consent) => {
  try {
    await saveConsent(payload)
    await saveConsentRequest(payload)

    pubSub.publish(SubscriptionMessage.CONSENT_GIVEN, {
      consentApproved: {
        accessToken: payload.accessToken,
      },
      consentRequestId: payload.consentRequestId,
    })
  } catch (e) {
    throw new Error(`Error in ConsentApproved: ${e}`)
  }
}

const onLoginApproved = (payload: Login) => {
  /**
   * TODO(@all): Should we save these?
   * */
  try {
    pubSub.publish(SubscriptionMessage.LOGIN_CONSENT_GIVEN, {
      loginApproved: {
        accessToken: payload.accessToken,
      },
      loginRequestId: payload.sessionId,
    })
  } catch (e) {
    throw new Error(`Error in LoginApproved: ${e}`)
  }
}

export { defaultRequest, onConsentApproved, onLoginApproved }
