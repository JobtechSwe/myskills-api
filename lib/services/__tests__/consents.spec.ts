import { defaultRequest, onConsentApproved } from '../consents'
import MockDate from 'mockdate'
import { saveConsent, saveConsentRequest } from '../db'
import pubSub from '../../adapters/pubsub'
jest.mock('../../config', () => ({ DOMAIN: 'myskills-api-domain' }))

jest.mock('../db', () => ({
  saveConsent: jest.fn(),
  saveConsentRequest: jest.fn(),
}))

jest.mock('../../adapters/pubsub', () => ({
  publish: jest.fn(),
}))

afterAll(MockDate.reset)

describe('#defaultRequest', () => {
  MockDate.set('2019-02-26T09:27:59.942Z')

  test('it creates a default consent', () => {
    expect(defaultRequest(3600)).toMatchSnapshot()
  })
})

describe('#onConsentApproved', () => {
  const consent = {
    consentId: '1337',
    consentRequestId: '1338',
    accessToken: '123',
  }

  test('saves consent', async () => {
    await onConsentApproved(consent)

    expect(saveConsent).toHaveBeenCalledWith(consent)
  })

  test('saves consent request', async () => {
    await onConsentApproved(consent)

    expect(saveConsentRequest).toHaveBeenCalledWith(consent)
  })

  test('publishes access token to apollo subscription/pubsub', async () => {
    const mockedpubSubData = {
      consentApproved: {
        accessToken: consent.accessToken,
      },
      consentRequestId: consent.consentRequestId,
    }
    await onConsentApproved(consent)

    expect(pubSub.publish).toHaveBeenCalledWith(
      'Consent given',
      mockedpubSubData
    )
  })

  test('throw when something goes wrong', async () => {
    ;(saveConsent as jest.Mock).mockRejectedValue('Err')

    await expect(onConsentApproved(consent)).rejects.toThrow('err')
  })
})
