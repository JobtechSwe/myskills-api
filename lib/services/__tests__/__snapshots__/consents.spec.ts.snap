// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#defaultRequest it creates a default consent 1`] = `
Object {
  "expiry": 1551176880,
  "scope": Array [
    Object {
      "area": 0,
      "description": "A list of your work experiences, educations, language proficiencies and so on that you have entered in the service.",
      "domain": "http://localhost:3000",
      "lawfulBasis": "CONSENT",
      "permissions": Array [
        "write",
      ],
      "purpose": "In order to create a CV using our website.",
    },
    Object {
      "area": 1,
      "description": "A list of your work experiences, educations, language proficiencies and so on that you have entered in the service.",
      "domain": "http://localhost:3000",
      "lawfulBasis": "CONSENT",
      "permissions": Array [
        "write",
      ],
      "purpose": "In order to create a CV using our website.",
    },
    Object {
      "area": 2,
      "description": "A list of your work experiences, educations, language proficiencies and so on that you have entered in the service.",
      "domain": "http://localhost:3000",
      "lawfulBasis": "CONSENT",
      "permissions": Array [
        "write",
      ],
      "purpose": "In order to create a CV using our website.",
    },
    Object {
      "area": 3,
      "description": "A list of your work experiences, educations, language proficiencies and so on that you have entered in the service.",
      "domain": "http://localhost:3000",
      "lawfulBasis": "CONSENT",
      "permissions": Array [
        "write",
      ],
      "purpose": "In order to create a CV using our website.",
    },
  ],
}
`;
